conclave-enclave/src/main/kotlin/com/r3/conclave/enclave/internal/NativeHostCallInterface.kt

```diff
@@ -0,0 +1,116 @@
+     */
+    private val threadStacks = ThreadLocal<Stack<StackFrame>>()
+    private val stack: Stack<StackFrame>
+        get() {
+            if (threadStacks.get() == null) {
+                threadStacks.set(Stack<StackFrame>())
+            }
+            return threadStacks.get()
+        }
+
+    private fun checkCallType(type: HostCallType) = check(type == stack.peek().callType) { "Call type mismatch" }
+
+    /**
+     * Internal method for initiating a host call with specific arguments.
+     * This should not be called directly, but instead by implementations in [HostCallInterface].
+     */
+    override fun executeCall(callType: HostCallType, parameterBuffer: ByteBuffer): ByteBuffer? {
+        stack.push(StackFrame(callType, null, null))
+
+        Native.jvmOcall(
+                callType.toShort(), CallInterfaceMessageType.CALL.toByte(), parameterBuffer.getAllBytes(avoidCopying = true))
+
+        val stackFrame = stack.pop()
+
+        stackFrame.exceptionBuffer?.let {
+            throw ThrowableSerialisation.deserialise(it)
+        }
+
+        return stackFrame.returnBuffer
+    }
+
+    /**
+     * Handle ecalls that originate from the host.
+     */
+    fun handleEcall(callTypeID: Short, ecallType: CallInterfaceMessageType, data: ByteBuffer) {
+        when (ecallType) {
+            CallInterfaceMessageType.CALL -> handleCallEcall(EnclaveCallType.fromShort(callTypeID), data)
+            CallInterfaceMessageType.RETURN -> handleReturnEcall(HostCallType.fromShort(callTypeID), data)
+            CallInterfaceMessageType.EXCEPTION -> handleExceptionEcall(HostCallType.fromShort(callTypeID), data)
+        }
+    }
+
+    /**
+     * In release mode we want exceptions propagated out of the enclave to be sanitised
+     * to reduce the likelihood of secrets being leaked from the enclave.
+     */
+    private fun sanitiseThrowable(throwable: Throwable): Throwable {
+        return when (throwable) {
+            is EnclaveStartException -> throwable
+            // Release enclaves still need to notify the host if they were unable to decrypt mail, but there's
+            // no need to include the message or stack trace in case any secrets can be inferred from them.
+            is MailDecryptionException -> MailDecryptionException()
+            else -> {
+                RuntimeException("Release enclave threw an exception which was swallowed to avoid leaking any secrets")
+            }
+        }
+    }
+
+    /**
+     * Handle call initiations from the host.
+     * This method propagates the call to the appropriate enclave side call handler, then sanitises, serialises and
+     * propagates any exceptions that occur. If a return value is produced, a reply message is sent back to the host.
+     */
+    private fun handleCallEcall(callType: EnclaveCallType, parameterBuffer: ByteBuffer) {
+        try {
+            acceptCall(callType, parameterBuffer)?.let {
```

Comment:
Is there any possibility that the buffer coming out of this call is empty? In that case we  seem to do nothing

Reply:
This is as intended. Null return values are signalled to the other side by not sending a return value or exception message. If the buffer is empty, then an ocall _will_ occur and an empty buffer will be sent. See `NativeHostCallinterface.kt:54`.

Reply:
As discussed, let's just add an explanation comment, feel free to close this once done